src/data/date-truncation.cpp

```diff
@@ -0,0 +1,72 @@
+#include <fstream>
+#include <string>
+#include <vector>
+#include <algorithm>
+#include "date-truncation.h"
+using namespace std;
+
+int get_days_for_month(int month, int year);
+bool isleap(int year);
+std::string TruncateDate(std::string date, std::string format) {
+    if (format.length() != 8 || date.length() != 8) {
+        return date;
+    }
+    for (int i = 0; char c: format) {
+        if (std::isdigit(c)) {
+            date.replace(i, 1, 1, c);
+        }
+        i++;
+    }
+    int year = std::stoi(date.substr(0, 4));
+    int month = std::stoi(date.substr(4, 2));
+    int day = std::stoi(date.substr(6, 2));
+    month = std::max(std::min(month, 12), 0);
+    day = std::max(std::min(day, get_days_for_month(month, year)), 0);
+    char truncated[24] = {0};
+    sprintf(truncated,"%d%02d%02d",year,month,day);
+    return truncated;
+}
+
+int get_days_for_month(int month, int year){
```

Comment:
How about we use an array to store the months? And with a check for February of course.

Reply:
Yea, that's a good idea. Save a couple ticks if we're lucky